test/matrix.lean

```diff
@@ -0,0 +1,129 @@
+run_cmd do let d ← get_dims ``(!![;;]),      guard $ d = (`(2), `(0))
+run_cmd do let d ← get_dims ``(!![,]);       guard $ d = (`(0), `(1))
+run_cmd do let d ← get_dims ``(!![,,]);      guard $ d = (`(0), `(2))
+run_cmd do let d ← get_dims ``(!![1]);       guard $ d = (`(1), `(1))
+run_cmd do let d ← get_dims ``(!![1,]);      guard $ d = (`(1), `(1))
+run_cmd do let d ← get_dims ``(!![1;]);      guard $ d = (`(1), `(1))
+run_cmd do let d ← get_dims ``(!![1,2;3,4]); guard $ d = (`(2), `(2))
+
+end dimensions
+
+run_cmd guard $ (!![1;2])       = of ![![1], ![2]]
+run_cmd guard $ (!![1,3])       = of ![![1,3]]
+run_cmd guard $ (!![1,2;3,4])   = of ![![1,2], ![3,4]]
+run_cmd guard $ (!![1,2;3,4;])  = of ![![1,2], ![3,4]]
+run_cmd guard $ (!![1,2,;3,4,]) = of ![![1,2], ![3,4]]
+
+example {a a' b b' c c' d d' : α} :
+  !![a, b; c, d] + !![a', b'; c', d'] = !![a + a', b + b'; c + c', d + d'] :=
+by simp
+
+example {a a' b b' c c' d d' : β} :
+  !![a, b; c, d] - !![a', b'; c', d'] = !![a - a', b - b'; c - c', d - d'] :=
+by simp
+
+example {a a' b b' c c' d d' : α} :
+  !![a, b; c, d] ⬝ !![a', b'; c', d'] =
+    !![a * a' + b * c', a * b' + b * d'; c * a' + d * c', c * b' + d * d'] :=
+by simp [-Equiv.Perm.coe_subsingleton]
+
+example {a b c d x y : α} :
+  mulVec !![a, b; c, d] ![x, y] = ![a * x + b * y, c * x + d * y] :=
+by simp
+
+example {a b c d : α} : submatrix !![a, b; c, d] ![1, 0] ![0] = !![c; a] :=
+by ext; simp
+
+example {a b c : α} : ![a, b, c] 0 = a := by simp
+example {a b c : α} : ![a, b, c] 1 = b := by simp
+example {a b c : α} : ![a, b, c] 2 = c := by simp
+
+example {a b c d : α} : ![a, b, c, d] 0 = a := by simp
+example {a b c d : α} : ![a, b, c, d] 1 = b := by simp
+example {a b c d : α} : ![a, b, c, d] 2 = c := by simp
+example {a b c d : α} : ![a, b, c, d] 3 = d := by simp
+example {a b c d : α} : ![a, b, c, d] 42 = c := by simp
+
+example {a b c d e : α} : ![a, b, c, d, e] 0 = a := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 1 = b := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 2 = c := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 3 = d := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 4 = e := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 5 = a := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 6 = b := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 7 = c := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 8 = d := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 9 = e := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 123 = d := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 123456789 = e := by simp
+
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 5 = f := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 7 = h := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 37 = f := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 99 = d := by simp
```

Comment:
If these tests fail we can just comment them out; they're not related to this PR